File: tools/sdks/flutter/android/src/main/kotlin/io/thunder/flutter/ThunderFlutterPlugin.kt
package io.thunder.flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class ThunderFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var handler: ThunderMethodHandler
    private val scope = CoroutineScope(Dispatchers.Main)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        handler = ThunderMethodHandler(binding.applicationContext)
        channel = MethodChannel(binding.binaryMessenger, "io.thunder/sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val args = call.arguments<Map<String, Any?>>() ?: emptyMap()
        scope.launch {
            handler.handle(call.method, args, result)
        }
    }
}

File: tools/sdks/flutter/ios/Classes/ThunderFlutterPlugin.swift
import Flutter
import UIKit
import Thunder

@objc public class ThunderFlutterPlugin: NSObject, FlutterPlugin {
    private let handler = ThunderMethodHandler()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "io.thunder/sdk", binaryMessenger: registrar.messenger())
        let instance = ThunderFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        Task {
            await handler.handle(method: call.method, args: args, result: result)
        }
    }
}
